src/preprocess/links.rs

```diff
@@ -0,0 +1,240 @@
+    }
+
+    replaced.push_str(&s[previous_end_index..]);
+    Ok(replaced)
+}
+
+#[derive(PartialOrd, PartialEq, Debug)]
+enum LinkType<'a> {
+    Escaped,
+    Include(PathBuf),
+    Playpen(PathBuf, Vec<&'a str>),
+}
+
+#[derive(PartialOrd, PartialEq, Debug)]
+struct Link<'a> {
+    start_index: usize,
+    end_index: usize,
+    link: LinkType<'a>,
+    link_text: &'a str,
+}
+
+impl<'a> Link<'a> {
+    fn from_capture(cap: Captures<'a>) -> Option<Link<'a>> {
+
+        let link_type = match (cap.get(0), cap.get(1), cap.get(2)) {
+            (_, Some(typ), Some(rest)) => {
+                let mut path_props = rest.as_str().split_whitespace();
+                let file_path = path_props.next().map(PathBuf::from);
+                let props: Vec<&str> = path_props.collect();
+
+                match (typ.as_str(), file_path) {
+                    ("include", Some(pth)) => Some(LinkType::Include(pth)),
+                    ("playpen", Some(pth)) => Some(LinkType::Playpen(pth, props)),
+                    _ => None,
+                }
+            },
+            (Some(mat), None, None) if mat.as_str().starts_with(ESCAPE_CHAR) => Some(LinkType::Escaped),
+            _ => None,
+        };
+
+        link_type.and_then(|lnk| {
+            cap.get(0)
+                .map(|mat| {
+                         Link {
+                             start_index: mat.start(),
+                             end_index: mat.end(),
+                             link: lnk,
+                             link_text: mat.as_str(),
+                         }
+                     })
+        })
+    }
+
+    fn render_with_path<P: AsRef<Path>>(&self, base: P) -> Result<String> {
+        let base = base.as_ref();
+        match self.link {
+            // omit the escape char
+            LinkType::Escaped => Ok((&self.link_text[1..]).to_owned()),
+            LinkType::Include(ref pat) => {
+                file_to_string(base.join(pat)).chain_err(|| format!("Could not read file for link {}", self.link_text))
+            },
+            LinkType::Playpen(ref pat, ref attrs) => {
+                let contents = file_to_string(base.join(pat))
+                    .chain_err(|| format!("Could not read file for link {}", self.link_text))?;
+                let ftype = if !attrs.is_empty() { "rust," } else { "rust" };
+                Ok(format!("```{}{}\n{}\n```\n", ftype, attrs.join(","), contents))
+            },
+        }
+    }
+}
+
+struct LinkIter<'a>(CaptureMatches<'a, 'a>);
+
+impl<'a> Iterator for LinkIter<'a> {
+    type Item = Link<'a>;
+    fn next(&mut self) -> Option<Link<'a>> {
+        for cap in &mut self.0 {
+            if let Some(inc) = Link::from_capture(cap) {
+                return Some(inc);
+            }
+        }
+        None
+    }
+}
+
+fn find_links(contents: &str) -> LinkIter {
+    // lazily compute following regex
+    // r"\\\{\{#.*\}\}|\{\{#([a-zA-Z0-9]+)\s*([a-zA-Z0-9_.\-:/\\\s]+)\}\}")?;
```

Comment:
> Most of the complexity comes from the backslashes, but still I am not sure it is very maintainable.

Right it might not be the easiest to understand or modify but we are expressing quite complex grammar here (and I don;t think the regex approach could be made much simpler).

We can either:

- keep the regex approach, add better comments and a set of comprehensive tests 
- add a separate parser code using some high level library or parsing and add a set of comprehensive tests ;)